### ISSUE TITLE
Port to Google Tensor G2/G3 on Pixel Phones

### ISSUE THREAD

**Issue by teaalltr** [ggerganov/llama.cpp#829]
It would be nice to use the TPU in those SoCs to improve speed

**Comment by Sovenok-Hacker** [comment]
> It would be nice to use the TPU in those SoCs to improve speed

Yes

**Comment by Titaniumtown** [comment]
Any progress on this? Would be really cool to take advantage of this feature!

**Comment by gjmulder** [comment]
Re-upping.

**Comment by ggerganov** [comment]
I don't think there is any API that allows us to use the TPU - is that correct?

**Comment by gjmulder** [comment]
> I don't think there is any API that allows us to use the TPU - is that correct?

I found this:

> Using graphics processing units (GPUs) to run your machine learning (ML) models can dramatically improve the performance and the user experience of your ML-enabled applications. On Android devices, you can enable GPU-accelerated execution of your models using a [delegate](https://www.tensorflow.org/lite/performance/delegates) and one of the following APIs:
> 
> Interpreter API - [guide](https://www.tensorflow.org/lite/android/delegates/gpu)
> Task library API - [guide](https://www.tensorflow.org/lite/android/delegates/gpu_task)
> Native (C/C++) API - this guide
> This guide covers advanced uses of the GPU delegate for the C API, C++ API, and use of quantized models. For more information about using the GPU delegate for TensorFlow Lite, including best practices and advanced techniques, see the [GPU delegates](https://www.tensorflow.org/lite/performance/gpu) page.

https://www.tensorflow.org/lite/android/delegates/gpu_native

**Comment by gjmulder** [comment]
Won't fix.

**Comment by phymbert** [comment]
for my curiosity using Vulkan backend is not enough ?

**Comment by gjmulder** [comment]
Vulkan probably isn't that efficient in terms of performance per Watt at int8 compared to the G2 and G3 TPUs:

https://ai-benchmark.com/ranking_processors

Benchmark looks to be using MobileBERT, so not necessarily applicable to llamas.

**Comment by teaalltr** [comment]
Not an expert here, but why it is not applicable to llama models if TF lite supports the backend?

**Comment by gjmulder** [comment]
Here's hoping UXL provides a standardised mobile API across all Android phones:

https://www.oneapi.io/spec/

**Comment by teaalltr** [comment]
> Not an expert here, but why it is not applicable to llama models if TF lite supports the backend?

Also, TF lite supports conversion from TF models https://www.tensorflow.org/lite/api_docs/python/tf/lite/TFLiteConverter
Dunno of any other technical limitation, please correct me if I'm wrong. Llama in TF exists, to the best of my knowledge

**Comment by Ananay-22** [comment]
I built Vulkan for a Pixel 7 pro, and it just works so much slower than a CPU build. using ngl=1 halves the speed from CPU inference and using ngl=2 onwards keeps halving it. (Measuring tokens per second in the llama-cli output)

I don't think performance per watt is the issue as mentioned here

> Vulkan probably isn't that efficient in terms of performance per Watt at int8 compared to the G2 and G3 TPUs:
> 
> https://ai-benchmark.com/ranking_processors
> 
> Benchmark looks to be using MobileBERT, so not necessarily applicable to llamas.

I was going to do a study to see if we can get away with using a more power consuming inference API. 

But I think the MALI GPU that the Pixel 7 Pro has is not efficient enough, and is probably struggling with such a large model (my theory is that there is actually a lot of data transfer happening at each inference layer for GPU access (I still am not familiar how the shared memory works as compared to dedicated VRAM) and so it just slows down. 

I am not sure how much the TPUs would impact performance - but it seems like we would first need to get the XLA build unless these TPU processors are completely different from the high end ones on kaggle

> It's possible, but I guess it requires substantial amount of engineering efforts. I've seen some Google folks make llama work through pytorch/xla https://pytorch.org/blog/path-achieve-low-inference-latency/. 

 _Originally posted by @ymwangg in [#3253](https://github.com/ggerganov/llama.cpp/issues/3253#issuecomment-1793230644)_

I am going to eventually check out if I can get this running on a pytorch build in the pixel 7 pro and see if that offers a speedup

**Comment by SuperPauly** [comment]
@ananay-22 How far did you get? I have a 8 Pro and would like to know if llama.cpp is the best way to run models.

**Comment by ananay-22** [comment]
Sorry - was part of a project I'm not working on currently so no progress from my side. 

CPU inference was "fine" from what I could remember. Atleast it was okay for what I was working on at the time - I'm guessing you have not had luck with it. But yes the issue with vulkan being significantly slower was quite bad and I just kept the model on the CPU.
